docs/design/2021-12-08-instance-scope.md

```diff
@@ -0,0 +1,187 @@
+@@ -1121,6 +1125,7 @@ func (s *session) GetGlobalSysVar(name string) (string, error) {
+ }
+ 
+ // SetGlobalSysVar implements GlobalVarAccessor.SetGlobalSysVar interface.
++// it is used for setting instance scope as well.
+ func (s *session) SetGlobalSysVar(name, value string) (err error) {
+        sv := variable.GetSysVar(name)
+        if sv == nil {
+@@ -1132,6 +1137,9 @@ func (s *session) SetGlobalSysVar(name, value string) (err error) {
+        if err = sv.SetGlobalFromHook(s.sessionVars, value, false); err != nil {
+                return err
+        }
++       if sv.HasInstanceScope() { // skip for INSTANCE scope
++               return nil
++       }
+        if sv.GlobalConfigName != "" {
+                domain.GetDomain(s).NotifyGlobalConfigChange(sv.GlobalConfigName, variable.OnOffToTrueFalse(value))
+        }
+@@ -1148,6 +1156,9 @@ func (s *session) SetGlobalSysVarOnly(name, value string) (err error) {
+        if err = sv.SetGlobalFromHook(s.sessionVars, value, true); err != nil {
+                return err
+        }
++       if !sv.HasInstanceScope() { // skip for INSTANCE scope
++               return nil
++       }
+        return s.replaceGlobalVariablesTableValue(context.TODO(), sv.Name, value)
+ }
+```
+
+A "non-native" `INSTANCE` variable can be changed to a native one as follows:
+
+```
++       {Scope: ScopeInstance, Name: TiDBGeneralLog, Value: BoolToOnOff(DefTiDBGeneralLog), Type: TypeBool, skipInit: true, SetGlobal: func(s *SessionVars, val string) error {
+                ProcessGeneralLog.Store(TiDBOptOn(val))
+                return nil
+-       }, GetSession: func(s *SessionVars) (string, error) {
++       }, GetGlobal: func(s *SessionVars) (string, error) {
+                return BoolToOnOff(ProcessGeneralLog.Load()), nil
+        }},
+```
+
+This introduces a compatibility issue, since users who have previously configured instance scope with `SET [SESSION] tidb_general_log = x` will receive an error because the scope is wrong. This can be fixed by adding a legacy mode to the sysvar framework which allows `INSTANCE` scoped variables to be set with either `SET GLOBAL` or `SET SESSION`:
+
+```sql
+SET GLOBAL tidb_enable_legacy_instance_scope = 1;
+```
+
+We can default to `TRUE` for now, but should revisit this in the future.
+
+### Stage 2: Mapping All Configuration File Settings to System Variables
+
+The second stage is to map all configuration file settings to system variable names. In MySQL any configuration file setting is also available as a system variable (even if it is read only), which makes it possible to get the configuration of the cluster with `SHOW GLOBAL VARIABLES`. Once this step is complete, it should also be possible to offer an `[instance]` section of the configuration file which accepts the system variable names (and finally consistent naming between the two systems). We can then modify the example configuration files to be based on the flat `[instance]` configuration:
```

Comment:
Does this mean that if I change the value of an system variable, the value in the configuration file will be changed as well ? or those var from config file is readonly?

Reply:
The default will be that we map variables as read-only, so we don't need to think about the specific cases (i.e. you can't easily change the socket or port). This helps us achieve completeness first, and then we can then evaluate which ones can be made dynamic.

When you modify an instance scoped system variable, it modifies in memory only (similar to `SET GLOBAL` in MySQL). This does not propose modifying or rewriting the config file (that would make it lose formatting, and it gets complex quickly).

Reply:
> Why do we offer system variables in the configuration file? To allow users to set the initial values through configuration files? Or to be compatible with MySQL?

The goal is to (1) have system variables *show* the superset of all configuration and (2) have unified names between configuration. These do technically improve MySQL compatibility, but I would say these are also good practices.

In the case of instance variables, they do not persist on restart so a configuration file is required to handle this case. In cockroachdb they don't have this though: node (their term for instance) is only supported as startup parameters, everything else is a cluster wide variable. But I think this is also an incompatible change.. by making every config setting available through systerm variables *and* settable in the config file using the system variable name (under `[instance]` heading) gives us a better transition state to unification.

> Initial values of some system variables are already configured in the configuration file, so they need to be deduplicated from showing global variables or the configuration file.

Because instance values do not persist, this behavior is straight forward. The configuration file values will be taken, and if not specified, compiled server defaults. This is also technically a MySQL compatible way for GLOBAL variables.